miden-lib/asm/eoa/basic.masm

```diff
@@ -1,32 +1,47 @@
 use.miden::sat::account
 use.miden::sat::tx
+use.std::crypto::dsa::rpo_falcon512
 
-#! Authenticate a transaction using Falcon
-#! WIP - FALCON VERIFICATION IS MISSING, NOT SURE IF HASHING IN CORRECT ORDER
+# CONSTANTS
+# =================================================================================================
+
+# Slot in account storage at which the public key is stored.
+const.PUBLIC_KEY_SLOT=0
+
+#! Authenticate a transaction using Falcon 
 #! Stack: []
 #! Output: []
 #!
-#! - ASSET is the non-fungible asset of interest.
-#! - tag is the tag to be included in the note.
-#! - RECIPIENT is the recipient of the note.
 export.auth_tx_rpo_falcon512
-    # update the nonce
-    push.1
-    exec.account::incr_nonce
-
-    # get commitments to consumed and created notes, new nonce, and ID
-    #exec.tx::get_output_notes_hash
-    #exec.tx::get_input_notes_hash
-    #exec.account::get_nonce push.0.0.0
-    #exec.account::get_id push.0.0.0
-
-    # compute the message to be signed
-    # M = h(output_notes_hash, h(input_notes_hash, h(0, 0, 0, id, 0, 0, 0, nonce)))
-    #hmerge hmerge hmerge
-
-    # get public key from account storage (assuming it is stored at index 0) and verify signature
-    #push.0
-    #exec.account::get_item
-    # [PUB_KEY, M]
-    # TODO: sign the message using FALCON
+    # Get commitments to consumed and created notes, current nonce, and ID
+    exec.tx::get_output_notes_hash
+    # => [OUTPUT_NOTES_HASH, ...]
+    
+    exec.tx::get_input_notes_hash
+    # => [INPUT_NOTES_HASH, OUTPUT_NOTES_HASH, ...]
+
+    exec.account::get_nonce push.0.0.0
+    # => [0, 0, 0, nonce, INPUT_NOTES_HASH, OUTPUT_NOTES_HASH, ...]
+
+    exec.account::get_id push.0.0.0
+    # => [0, 0, 0, account_id, 0, 0, 0, nonce, INPUT_NOTES_HASH, OUTPUT_NOTES_HASH, ...]
+
+    # Compute the message to be signed
+    # M = h(OUTPUT_NOTES_HASH, h(INPUT_NOTES_HASH, h(0, 0, 0, account_id, 0, 0, 0, nonce)))
+    hmerge hmerge hmerge
```

Comment:
question: Is there a reason we use 2 to 1 hashing here instead of using a sequential hash over all elements which would be more efficient?

Reply:
Isn't the 2-to-1 hash more efficient? 

https://0xpolygonmiden.github.io/miden-vm/user_docs/assembly/cryptographic_operations.html

Or maybe I don't get it. Which instruction would you use?

Reply:
The `hperm` operation for sequential hashing.  It would look something like this (please review yourself and adjust accordingly to ensure correctness):

```
# prepare hasher capacity elements
padw 

# prepare first hasher rate word
exec.account::get_nonce push.0.0.0

# prepare second hasher rate word
exec.account::get_id push.0.0.0

# apply first permutation
hperm

# drop rate elements
dropw dropw

# prepare third hasher rate word
exec.tx::get_input_notes_hash

# prepare fourth hasher rate word
exec.tx::get_output_notes_hash

# apply second hasher permutation
hperm

# extract the digest (message)
dropw swapw dropw
```

This will give a different result to the `hmerge` method but it is a more efficient way of computing a message.

Reply:
`hmerge` was used as it was the easiest way to construct how the message M was specified. 

`M = h(OUTPUT_NOTES_HASH, h(INPUT_NOTES_HASH, h(0, 0, 0, account_id, 0, 0, 0, nonce)))`

However, if the other method is more efficient. Should I open an issue and we discuss this?

Reply:
Yes please lets leave it as it is and open an issue to discuss this.